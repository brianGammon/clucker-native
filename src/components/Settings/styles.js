/* @flow */
import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  rowContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  profileContainer: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  profileIcon: {
    fontSize: 40,
  },
  profileInfo: {
    paddingLeft: 10,
  },
  sectionLabel: {
    fontSize: 24,
    fontWeight: 'bold',
  },
});
