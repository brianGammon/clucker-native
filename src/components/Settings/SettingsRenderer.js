/* @flow */
import * as React from 'react';
import {
  Container, Content, View, Text, Button, Icon,
} from 'native-base';
import Header from '../Header';
import CommonLabel from '../CommonLabel';
import { type User } from '../../types';
import styles from './styles';

type Props = {
  user: User,
  handleSignOut: () => void,
};

const SettingsRenderer = ({ handleSignOut, user }: Props) => (
  <Container>
    <Header title="Settings" />
    <Content padder>
      <Text style={styles.sectionLabel}>Account</Text>
      <View padder style={styles.rowContainer}>
        <View style={styles.profileContainer}>
          <Icon style={styles.profileIcon} active name="person" />
          <View style={styles.profileInfo}>
            <CommonLabel text="Logged In:" />
            <Text>{user ? user.email : 'unknown user'}</Text>
          </View>
        </View>

        <Button transparent onPress={handleSignOut}>
          <Text>Sign Out</Text>
        </Button>
      </View>
    </Content>
  </Container>
);
export default SettingsRenderer;
