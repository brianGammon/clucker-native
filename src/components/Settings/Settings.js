/* @flow */
import * as React from 'react';
import { Alert } from 'react-native';
import { connect } from 'react-redux';
import SettingsRenderer from './SettingsRenderer';
import { type Navigation, type Flock, type UserSettings } from '../../types';
import { firebaseUpdateRequested } from '../../redux/actions';
import { actionTypes, metaTypes } from '../../redux/constants';

type Props = {
  navigation: Navigation,
  flocks: {
    [flockId: string]: Flock,
  },
  userSettings: UserSettings,
  userId: string,
  signOut: () => void,
  saveUserSettings: (userId: string, userSettings: UserSettings) => void,
  joinFlock: (userId: string, flockId: string) => void,
  joinError: string,
  joinInProgess: boolean,
};

type State = {
  joinForm: {
    value: string,
    error: string,
    touched: boolean,
  },
  // addForm: {
  //   value: string,
  //   error: string,
  //   touched: boolean,
  // },
};

const formDefaultValues = {
  value: '',
  error: '',
  touched: false,
};

class Settings extends React.Component<Props, State> {
  state = {
    joinForm: {
      ...formDefaultValues,
    },
    // addForm: {
    //   value: '',
    //   error: '',
    //   touched: '',
    // }
  };

  componentDidUpdate(prevProps) {
    const {
      userSettings: { currentFlockId },
      flocks,
      joinInProgess,
      joinError,
    } = this.props;

    if (
      currentFlockId
      && currentFlockId !== prevProps.userSettings.currentFlockId
    ) {
      Alert.alert(`"${flocks[currentFlockId].name}" is now active`, null, [
        { text: 'OK' },
      ]);
    }
    if (prevProps.joinInProgess && !joinInProgess && joinError === '') {
      this.resetForm('joinForm');
    }
  }

  resetForm = (form: string) => {
    this.setState({ [form]: { ...formDefaultValues } });
  };

  handleChangeText = (form: string, text: string) => {
    const { [form]: formState } = this.state;
    this.setState({ [form]: { ...formState, value: text } });
  };

  handleSelectFlock = (flockId) => {
    const { userId, userSettings, saveUserSettings } = this.props;
    const newUserSettings = { ...userSettings, currentFlockId: flockId };
    saveUserSettings(userId, newUserSettings);
  };

  handleJoinFlock = () => {
    const { userId, joinFlock } = this.props;
    const {
      joinForm: { value },
    } = this.state;
    joinFlock(userId, value);
  };

  handleSignOut = () => {
    const { navigation, signOut } = this.props;
    signOut();
    navigation.navigate('SignedOut');
  };

  render() {
    const {
      flocks,
      userId,
      navigation,
      userSettings: { currentFlockId },
      joinError,
    } = this.props;
    return (
      <SettingsRenderer
        flocks={flocks}
        currentFlockId={currentFlockId}
        userId={userId}
        handleSignOut={this.handleSignOut}
        handleSelectFlock={this.handleSelectFlock}
        navigation={navigation}
        handleJoinFlock={this.handleJoinFlock}
        handleChangeText={this.handleChangeText}
        {...this.state}
        joinError={joinError}
      />
    );
  }
}

const mapStateToProps = ({
  flocks,
  userSettings,
  auth: { user },
  joinForm: { error: joinError, inProgress: joinInProgess },
}) => ({
  flocks: flocks.data,
  userSettings: userSettings.data,
  userId: user ? user.uid : '',
  joinError,
  joinInProgess,
});

const mapDispatchToProps = dispatch => ({
  signOut: () => dispatch({ type: actionTypes.SIGN_OUT_REQUESTED }),
  saveUserSettings: (userId: string, userSettings: UserSettings) => dispatch(
    firebaseUpdateRequested({ userId, userSettings }, metaTypes.userSettings),
  ),
  joinFlock: (userId: string, flockId: string) => dispatch({
    type: actionTypes.JOIN_FLOCK_REQUESTED,
    payload: { userId, flockId },
  }),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Settings);
