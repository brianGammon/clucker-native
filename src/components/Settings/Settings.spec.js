import React from 'react';
import Settings from '.';

describe('Settings component:', () => {
  test('Should render', () => {
    const wrapper = shallow(<Settings />);
    expect(wrapper).toMatchSnapshot();
  });
});
