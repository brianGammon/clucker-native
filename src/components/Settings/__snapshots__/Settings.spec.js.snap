// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings component: Should render 1`] = `
<SettingsRenderer
  currentFlockId="flock1"
  flocks={
    Object {
      "flock1": Object {
        "name": "Test Flock 1",
        "ownedBy": "user1",
      },
      "flock2": Object {
        "name": "Test Flock 2",
        "ownedBy": "user1",
      },
    }
  }
  handleChangeText={[Function]}
  handleJoinFlock={[Function]}
  handleSelectFlock={[Function]}
  handleSignOut={[Function]}
  joinError=""
  joinForm={
    Object {
      "error": "",
      "touched": false,
      "value": "",
    }
  }
  userId="user1"
/>
`;
